src/disposer/AsyncDisposer.ts

```diff
@@ -0,0 +1,65 @@
+import { catchAllPromises } from './catchAllPromises';
+import { IAsyncDisposable } from './types';
+
+const listeners = new WeakMap<DisposeFn, Set<DisposeFn>>();
```

Comment:
Hm. This is tad risky IMHO.
What if the same dispose fn is added to more than a single instance of AsyncDisposer?
At the very least we should throw when that happens.